presto-jdbc/src/main/java/io/prestosql/jdbc/PrestoPreparedStatement.java

```diff
@@ -858,4 +876,69 @@ private static String typedNull(String prestoType)
     {
         return format("CAST(NULL AS %s)", prestoType);
     }
+
+    private static List<ColumnInfo> getDescribeOutputColumnInfoList(ResultSet resultSet)
+            throws SQLException
+    {
+        ImmutableList.Builder<ColumnInfo> list = ImmutableList.builder();
+        while (resultSet.next()) {
+            String columnName = resultSet.getString("Column Name");
+            String catalog = resultSet.getString("Catalog");
+            String schema = resultSet.getString("Schema");
+            String table = resultSet.getString("Table");
+            ClientTypeSignature clientTypeSignature = getClientTypeSignatureFromTypeString(resultSet.getString("Type"));
+            ColumnInfo.Builder builder = new ColumnInfo.Builder()
+                    .setColumnName(columnName)
+                    .setColumnLabel(columnName)
+                    .setCatalogName(catalog)
+                    .setSchemaName(schema)
+                    .setTableName(table)
+                    .setColumnTypeSignature(clientTypeSignature)
+                    .setNullable(ColumnInfo.Nullable.UNKNOWN);
+            setTypeInfo(builder, clientTypeSignature);
+            list.add(builder.build());
+        }
+        return list.build();
+    }
+
+    @VisibleForTesting
+    static ClientTypeSignature getClientTypeSignatureFromTypeString(String type)
+    {
+        String topLevelType;
+        List<ClientTypeSignatureParameter> arguments = new ArrayList<>();
+        Matcher topLevelMatcher = TOP_LEVEL_TYPE_PATTERN.matcher(type);
+        if (topLevelMatcher.matches()) {
+            topLevelType = topLevelMatcher.group(1);
+            String typeParameters = topLevelMatcher.group(2);
+            if (topLevelType.equals("decimal")) {
+                List<String> precisionAndScale = Splitter.on(',').splitToList(typeParameters);
+                checkArgument(precisionAndScale.size() == 2, "Invalid decimal parameters: %s", typeParameters);
+                arguments.add(ClientTypeSignatureParameter.ofLong(parseLong(precisionAndScale.get(0))));
+                arguments.add(ClientTypeSignatureParameter.ofLong(parseLong(precisionAndScale.get(1))));
+            }
+            else if (topLevelType.equals("char") || topLevelType.equals("varchar")) {
+                long precision = parseLong(typeParameters);
+                arguments.add(ClientTypeSignatureParameter.ofLong(precision));
+            }
```

Comment:
there should be a comment explaining why we discard type parameters for array, map, row

Reply:
@findepi Thanks! I added a TODO as a comment below base on the discussion we had before (which led us to a point that we only need top-level type currently): https://prestosql.slack.com/archives/CP1MUNEUX/p1570293305067000?thread_ts=1570147672.028700&cid=CP1MUNEUX